test/e2e/e2e_test.go

```diff
@@ -0,0 +1,171 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	jobset "sigs.k8s.io/jobset/api/v1alpha1"
+	"sigs.k8s.io/jobset/pkg/util/testing"
+)
+
+var _ = ginkgo.Describe("JobSet", func() {
+
+	// Each test runs in a separate namespace.
+	var ns *corev1.Namespace
+
+	ginkgo.BeforeEach(func() {
+		// Create test namespace before each test.
+		ns = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "e2e-",
+			},
+		}
+		gomega.Expect(k8sClient.Create(ctx, ns)).To(gomega.Succeed())
+
+		// Wait for namespace to exist before proceeding with test.
+		gomega.Eventually(func() bool {
+			err := k8sClient.Get(ctx, types.NamespacedName{Namespace: ns.Namespace, Name: ns.Name}, ns)
+			if err != nil {
+				return false
+			}
+			return true
+		}, timeout, interval).Should(gomega.BeTrue())
+	})
+
+	ginkgo.AfterEach(func() {
+		// Delete test namespace after each test.
+		gomega.Expect(k8sClient.Delete(ctx, ns)).To(gomega.Succeed())
+	})
+
+	type testCase struct {
+		makeJobSet      func(*corev1.Namespace) *testing.JobSetWrapper
+		expectCondition jobset.JobSetConditionType
+	}
+
+	ginkgo.DescribeTable("",
+		func(tc *testCase) {
+			ctx := context.Background()
+
+			// Create JobSet.
+			ginkgo.By("creating jobset")
+			js := tc.makeJobSet(ns).Obj()
+
+			// Verify jobset created successfully.
+			ginkgo.By("checking that jobset creation succeeds")
+			gomega.Expect(k8sClient.Create(ctx, js)).Should(gomega.Succeed())
+
+			// We'll need to retry getting this newly created jobset, given that creation may not immediately happen.
+			gomega.Eventually(k8sClient.Get(ctx, types.NamespacedName{Name: js.Name, Namespace: js.Namespace}, &jobset.JobSet{}), timeout, interval).Should(gomega.Succeed())
+
+			ginkgo.By("checking all jobs were created successfully")
+			gomega.Eventually(checkNumJobs, timeout, interval).WithArguments(ctx, js).Should(gomega.Equal(numExpectedJobs(js)))
+
+			// Check jobset status if specified.
+			ginkgo.By("checking jobset condition")
+			gomega.Eventually(checkJobSetStatus, timeout, interval).WithArguments(js, tc.expectCondition).Should(gomega.Equal(true))
+		},
+		ginkgo.Entry("pods can reach each other via hostname when DNS hostnames are enabled", &testCase{
+			makeJobSet:      pingTestJobSet,
+			expectCondition: jobset.JobSetCompleted,
+		}),
+	) // end of DescribeTable
+}) // end of Describe
+
+func checkJobSetStatus(js *jobset.JobSet, condition jobset.JobSetConditionType) (bool, error) {
```

Comment:
This and the two functions below should be moved to a common library shared with the integration tests; may be under test/util/util.go (I don't like the name, but I can't think of a better one :()